Chart: Rework `NetworkPolicies`.

Gacko opened the pull request:
<!--- Provide a general summary of your changes in the Title above --->
<!--- Please don't @-mention people in PR or commit messages (do so in an additional comment). --->

## What this PR does / why we need it:
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

This PR reworks the existing controller network policy which was added for the admission webhook to restrict access on all known ports instead of allowing any and adds a network policy for the default backend.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] CVE Report (Scanner found CVE and adding report)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation only

## Which issue/s this PR fixes
<!--
(optional, in `fixes #<issue number>` format, will close that issue when PR gets merged):

fixes #
-->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/kubernetes/ingress-nginx/blob/main/CONTRIBUTING.md) guide
- [ ] I have added unit and/or e2e tests to cover my changes.
- [x] All new and existing tests passed.
- [x] Added Release Notes.

## Does my pull request need a release note?
Any user-visible or operator-visible change qualifies for a release note. This could be a:

- CLI change
- API change
- UI change
- configuration schema change
- behavioral change
- change in non-functional attributes such as efficiency or availability, availability of a new platform
- a warning about a deprecation
- fix of a previous Known Issue
- fix of a vulnerability (CVE)

No release notes are required for changes to the following:

- Tests
- Build infrastructure
- Fixes for unreleased bugs

For more tips on writing good release notes, check out the [Release Notes Handbook](https://github.com/kubernetes/sig-release/tree/master/release-team/role-handbooks/release-notes)

<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Chart: Rework `NetworkPolicies`.
```

Gacko commented:
@strongjz Would you mind reviewing and eventually merging this? :)

tao12345666333 commented:
/assign

tao12345666333 commented:
Can you provide more explanation for why this change is necessary, given that it is a disruptive modification? Thanks

Gacko commented:
Sure!

Right now there's a `NetworkPolicy` called `{{ include "ingress-nginx.fullname" . }}-webhooks-allow` and only created if the admission webhook (`controller.admissionWebhooks.enabled`) and its network policy (`controller.admissionWebhooks.networkPolicyEnabled`) are enabled. This would make sense if it would only allow access to the actual webhook ports, but actually it allows access to any port of the controller ports.

I first of all I just renamed the manifest and made it what it's actually doing plus not binding it to the admission webhook configuration but to the controller configuration and limit access to the ports really in use. One can still disable the access to the admission webhook port by simply disabling `controller.admissionWebhooks.enabled`:

https://github.com/kubernetes/ingress-nginx/pull/9634/files#diff-07a869fb92716cdbb5321f9be8e26f65ade88f5ab6ed5bf073b690cf00548d3cR31

So yeah, you're right: This is breaking because I removed `controller.admissionWebhooks.networkPolicyEnabled` or at least replaced it by `controller.networkPolicy.enabled` (kind of).

Additionally I created a similar network policy for the default backend, which did not exist before, and reworked the existing network policy for the webhook patch jobs to align it to the other webhook patch job resources.

If one requires `NetworkPolicies` in his/hers environment, they enabled `controller.admissionWebhooks.networkPolicyEnabled` before and will now also enable the newly introduced and in the release notes mentioned `controller.networkPolicy.enabled` to make the whole `ingress-nginx` work in their setup, which in the end results in the same.

So from my point of view: Yeah, it's breaking, but right now in environments which require `NetworkPolicies` to make `ingress-nginx` work, it only works when you enable `controller.admissionWebhooks.networkPolicyEnabled` - which is way more confusing - or add a `NetworkPolicy` on your own. The latter is introduced in this PR. So I'm just making things work straight forward and clearer.

btw: This all is totally useless and non-breaking in environments not using restrictive / deny-by-default `NetworkPolices` aka it only applies to those restrictive environments and users of the latter have way more problems right now than without my change.

tao12345666333 commented:
Thank you for continuously updating this PR. I need to do some verification on it. You don't need to update it anymore before I come back. I will leave comments when I finish the review.

Gacko commented:
Hey @tao12345666333! Have you been able to review this PR? Is there anything I can do to get it merged soon?

Gacko commented:
I opened https://github.com/kubernetes/ingress-nginx/pull/10238 in favor of this PR since the base repository changed.
